pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+package chartsync
+
+import (
+	"context"
+	"errors"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/weaveworks/flux/git"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/fluxcd/helm-operator/pkg/apis/helm.fluxcd.io/v1"
+	lister "github.com/fluxcd/helm-operator/pkg/client/listers/helm.fluxcd.io/v1"
+)
+
+// Various (final) errors.
+var (
+	ErrReleasesForMirror = errors.New("failed to get HelmRelease resources for mirror")
+	ErrNoMirror          = errors.New("no existing git mirror found")
+	ErrMirrorSync        = errors.New("failed syncing git mirror")
+)
+
+// ReleaseQueue is an add-only `workqueue.RateLimitingInterface`
+type ReleaseQueue interface {
+	AddRateLimited(item interface{})
+}
+
+// GitConfig holds the configuration for git operations.
+type GitConfig struct {
+	GitTimeout      time.Duration
+	GitPollInterval time.Duration
+}
+
+// GitChartDir holds a `git.Export` of a chart repository,
+// and the commit hash of the HEAD of this export.
+type GitChartDir struct {
+	*git.Export
+	Head string
+}
+
+func (d GitChartDir) ChartPath(relativePath string) string {
+	if d.Export != nil {
+		return filepath.Join(d.Dir(), relativePath)
+	}
+	return ""
+}
+
+// GitChartSync syncs `sourceRef`s with their mirrors, and queues
+// updates for `v1.HelmRelease`s the sync changes are relevant for.
+type GitChartSync struct {
+	logger log.Logger
+	config GitConfig
+
+	lister lister.HelmReleaseLister
+
+	mirrors *git.Mirrors
+
+	sourcesMu sync.RWMutex
+	sources   map[string]sourceRef
```

Comment:
I think it's worth adding a comment about the map being indexed by the `ResourceID` of `HelmReleases` and probably also renaming it to `releaseSources` or `releaseSourcesByID`

Reply:
I felt that changing the name to `releaseSourcesByID` made the comment itself obsolete.